service/storage_service.cc

```diff
@@ -4412,7 +4412,54 @@ future<> storage_service::move(token new_token) {
 
 future<std::vector<storage_service::token_range_endpoints>>
 storage_service::describe_ring(const sstring& keyspace, bool include_only_local_dc) const {
-    return locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+    if (_db.local().find_keyspace(keyspace).get_replication_strategy().uses_tablets()) {
+        throw std::runtime_error(format("The keyspace {} has tablet table. Query describe_ring with the table parameter!", keyspace));
+    }
+    co_return co_await locator::describe_ring(_db.local(), _gossiper, keyspace, include_only_local_dc);
+}
+
+future<std::vector<dht::token_range_endpoints>>
+storage_service::describe_ring_for_table(const sstring& keyspace_name, const sstring& table_name) const {
+    slogger.debug("describe_ring for table {}.{}", keyspace_name, table_name);
+    auto& t = _db.local().find_column_family(keyspace_name, table_name);
+    if (!t.uses_tablets()) {
+        auto ranges = co_await describe_ring(keyspace_name);
+        co_return ranges;
+    }
+    table_id tid = t.schema()->id();
+    auto erm = t.get_effective_replication_map();
+    auto& tmap = erm->get_token_metadata_ptr()->tablets().get_tablet_map(tid);
+    const auto& topology = erm->get_topology();
+    std::unordered_map<locator::tablet_id, locator::tablet_info> maps;
+    co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) {
```

Comment:
IIUC, `tmap.for_each_tablet()` does not invoke the given function on the same tablet for multiple times, as the tablets in `tablet_map::_tablets` are unique. neither do we need to index in the map for traversing in it. so probably, we don't need to use an `unordered_map<>` for holding the tuple of `tablet_id` and `tablet_info`, a vector of pair or a tuple would suffice.